onnxruntime/core/session/inference_session.cc

```diff
@@ -1733,59 +1746,115 @@ common::Status InferenceSession::Run(IOBinding& io_binding) {
 }
 
 #ifdef ENABLE_TRAINING
-common::Status InferenceSession::RunInBackgroundAndWaitForYield(RunOptions& run_options, IOBinding& io_binding,
-                                                                std::vector<OrtValue>& user_outputs) {
-  const int64_t main_thread_event_id = 0;
-  onnxruntime::contrib::OrtEventPool::GetInstance().ResetEvent(0);
-
-  task_.terminate_flag_ = &(run_options.terminate);
-  task_.bg_thread_promise_ = std::promise<Status>();
-  task_.bg_thread_future_ = task_.bg_thread_promise_.get_future();
-  task_.bg_thread_ = std::thread([&](std::promise<common::Status> result_promise) {
-    common::Status s = Run(run_options, io_binding.GetInputNames(), io_binding.GetInputs(), io_binding.GetOutputNames(),
-                           &io_binding.GetOutputs(), &io_binding.GetOutputsDeviceInfo());
-
-    result_promise.set_value(s);
-
-    // signal main thread for background thread completion
-    const int64_t main_thread_event_id = 0;
-    onnxruntime::contrib::OrtEventPool::GetInstance().SignalEvent(main_thread_event_id);
+common::Status InferenceSession::RunInBackgroundAndWaitForYield(const RunOptions& run_options, IOBinding& io_binding,
+                                                                std::vector<OrtValue>& user_outputs, int64_t& run_id) {
+  if (terminating_runs_.load()) {
```

Comment:
IIUC this check (and flag) shouldn't be necessary, and we should never take the true branch here. Since we're only setting the flag to true in the destructor, in order for `terminating_runs_` to be true here, we'd have to call this method either after or during a call to the destructor. And destroying an object while calling another method on it should be undefined behavior.

Reply:
ok.